sycl/source/detail/scheduler/graph_builder.cpp

```diff
@@ -1387,6 +1481,186 @@ Command *Scheduler::GraphBuilder::connectDepEvent(
   return ConnectCmd;
 }
 
+void Scheduler::GraphBuilder::startFusion(QueueImplPtr Queue) {
+  auto QUniqueID = std::hash<QueueImplPtr>()(Queue);
+  if (isInFusionMode(QUniqueID)) {
+    throw sycl::exception{sycl::make_error_code(sycl::errc::invalid),
+                          "Queue already in fusion mode"};
+  }
+  auto OldFusionCmd = findFusionList(QUniqueID);
+  if (OldFusionCmd != MFusionMap.end()) {
+    // If fusion was used on this queue previously, the old fusion command might
+    // still be around to make sure that even after
+    // cancellation of the fusion due to synchronization, complete_fusion is
+    // still able to return a valid event.
+    OldFusionCmd->second->setFusionStatus(
+        KernelFusionCommand::FusionStatus::DELETED);
+    cleanupCommand(OldFusionCmd->second.release());
+    MFusionMap.erase(OldFusionCmd);
+  }
+  MFusionMap.emplace(QUniqueID, std::make_unique<KernelFusionCommand>(Queue));
+}
+
+void Scheduler::GraphBuilder::removeNodeFromGraph(
+    Command *Node, std::vector<Command *> &ToEnqueue) {
+  // Remove the placeholder command as leaf of all its requirements and from the
+  // user list of all its dependencies.
+  for (auto &Dep : Node->MDeps) {
+    auto AccessMode = Dep.MDepRequirement->MAccessMode;
+    auto *Record = getMemObjRecord(Dep.MDepRequirement->MSYCLMemObj);
+
+    Node->MLeafCounter -= Record->MReadLeaves.remove(Node);
+    Node->MLeafCounter -= Record->MWriteLeaves.remove(Node);
+    // If the placeholder had a write-requirement on this record, we need to
+    // restore the previous leaves.
+    if (AccessMode != access::mode::read) {
```

Comment:
This looks somewhat worrying. Are we always able to restore the correct leaf commands here? Can there be a scenario where the previous leaf command has already been destroyed at this point (e.g. once it stops being a leaf assuming it's already been enqueued) and we just lose the dependency when adding the fused kernel?

Reply:
I'm not sure, I might be getting something wrong. Let's say there's the following graph (assume all deps have write access): 
`Alloca <- CG`. We then start fusion: `Alloca <- CG <- Fusion list CG A <- Fusion list CG B`, CG is removed from leaves and cleaned up during this process, but isn't finished, so the graph looks like this: `Alloca <- Fusion list CG A <- Fusion list CG B`. Since the event associated with CG has already been added to the list of event dependencies of `Fusion list CG A` when building the graph, we'll honor that dependency if the fusion is cancelled and placeholder CGs are enqueued as normal despite this cleanup.

Would the same hold true if we complete the fusion? We remove the fusion list CGs first while restoring the leaves and then add the fusion cmd to the graph. It seems to me like the only dependency for the fusion command in this scenario would be `Alloca` while `CG` would be lost. Am I missing some way this dependency would be preserved here?

Reply:
So far, I assumed that if `CG` gets to the stage where it is cleaned up (destroyed) and wouldn't be available for restoring as a leave through `Fusion list CG A`'s `MDep`, it would already have completed execution. 
 
However, if `CG` can proceed to clean-up when it has been enqueued, but not necessarily executed yet, the scenario you pointed out might indeed be problematic. 

I have pushed a new commit to address this problem by inheriting all dependencies (apart from the ones on other commands in the fusion list) from the input commands in the fusion list. So, in your example, when we call `complete_fusion`, the new command for the fused kernel should inherit the dependency on `CG` from the input command `Fusion list CG A`, so we have the correct dependencies set up.

Reply:
Thanks for the feedback @sergey-semenov!

> Could you please add a test for this scenario as well?

This will not be possible as part of this PR. The logic to create a fused kernel from the kernels in the fusion list is only part of a separate, later PR (draft for that PR: https://github.com/sommerlukas/llvm/pull/1). 

Currently, `complete_fusion` always falls back to `cancel_fusion`, see the `TODO` in `complete_fusion` at `graph_builder.cpp:1548`. 

If it is ok with you, I would suggest to merge this PR and I will add a unit test for this scenario to the next PR, which introduces the necessary JIT compilation logic.